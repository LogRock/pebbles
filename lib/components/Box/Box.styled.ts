import styled, { css } from "styled-components";
import { BoxProps } from "./Box.types";

export const StyledBox = styled.div<Required<BoxProps>>`
  ${({ theme, ...props }) => css`
    display: flex;
    flex-direction: ${props.direction};
    flex-wrap: ${props.noWrap ? "no-wrap" : "wrap"};
    align-items: ${props.align};
    justify-content: ${props.justify};
    margin: ${theme.spacings[props.margin]};
    padding: ${theme.spacings[props.padding]};
    column-gap: ${theme.spacings[props.colGap]};

    border-radius: ${theme.spacings[props.borderRadius]};

    background-color: ${theme.colors[props.backgroundColor][
      props.backgroundShade
    ]};
    box-shadow: ${theme.shadows[props.shadow]};
    row-gap: ${theme.spacings[props.rowGap]};
  `}
`;
