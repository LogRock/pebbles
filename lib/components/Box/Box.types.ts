import * as CSS from "csstype";

import {
  colorTokens,
  shadeTokens,
  shadowTokens,
  spacingTokens,
  widthTokens,
} from "../../types/tokens";

export type BoxProps = {
  margin?: spacingTokens;
  padding?: spacingTokens;
  backgroundColor?: colorTokens;
  backgroundShade?: shadeTokens;
  borderRadius?: spacingTokens;
  shadow?: shadowTokens;
  direction?: CSS.Property.FlexDirection;
  rowGap?: spacingTokens;
  colGap?: spacingTokens;
  align?: CSS.Property.AlignItems;
  justify?: CSS.Property.JustifyContent;
  noWrap?: boolean;
  width?: widthTokens;
  grow?: boolean;
  shrink?: boolean;
  spaceAfter?: spacingTokens;
  borderColor?: colorTokens;
  borderShade?: shadeTokens;
  borderWidth?: spacingTokens;
  display?: CSS.Property.Display;
  gridColumns?: CSS.Property.GridTemplateColumns;
  gridRows?: CSS.Property.GridTemplateRows;
  gridArea?: CSS.Property.GridArea;
};
