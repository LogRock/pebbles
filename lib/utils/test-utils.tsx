/* eslint-disable import/export */
import React from "react";
import { cleanup, render } from "@testing-library/react";
import { afterEach } from "vitest";
import { ThemeProvider } from "styled-components";
import { light as theme } from "../theme";

afterEach(() => {
  cleanup();
});

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => (
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    ),
    ...options,
  });

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
// override render export
export { customRender as render };
