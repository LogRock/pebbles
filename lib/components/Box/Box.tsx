import styled, { css } from "styled-components";
import { widthTokens } from "../../types/tokens";
import { BoxProps } from "./Box.types";
import * as CSS from "csstype";

const widthMap: Record<widthTokens, CSS.Property.Width> = {
  "1/12": "8.33333333333%",
  "2/12": "16.6666666667%",
  "3/12": "25%",
  "4/12": "33.3333333333%",
  "5/12": "41.6666666667%",
  "6/12": "50%",
  "7/12": "58.3333333333%",
  "8/12": "66.6666666667%",
  "9/12": "75%",
  "10/12": "83.3333333333%",
  "11/12": "91.6666666667%",
  "12/12": "100%",
};

const Box = styled.div<BoxProps>`
  ${({
    theme,
    align = "center",
    backgroundColor,
    backgroundShade = "500",
    borderRadius = "zero",
    colGap = "zero",
    direction = "column",
    justify = "flex-start",
    margin = "zero",
    padding = "zero",
    rowGap = "zero",
    shadow = "none",
    noWrap = false,
    width = "12/12",
    grow = false,
    shrink = false,
    spaceAfter = "zero",
    borderColor = "neutral",
    borderShade = "200",
    borderWidth = "zero",
    display = "flex",
    gridColumns = "unset",
    gridRows = "unset",
    gridArea = "unset",
  }) => css`
    display: ${display};
    box-sizing: border-box;
    grid-area: ${gridArea};
    grid-template-columns: ${gridColumns};
    grid-template-rows: ${gridRows};
    flex-direction: ${direction};
    flex-grow: ${grow ? "1" : "0"};
    flex-shrink: ${shrink ? "1" : "0"};
    flex-wrap: ${noWrap ? "no-wrap" : "wrap"};
    align-items: ${align};
    justify-content: ${justify};
    width: ${widthMap[width]};
    margin: ${theme.spacings[margin]};
    ${spaceAfter
      ? css`
          margin-bottom: ${theme.spacings[spaceAfter]};
        `
      : ""}
    padding: ${theme.spacings[padding]};
    column-gap: ${theme.spacings[colGap]};

    border-width: ${theme.spacings[borderWidth]};
    border-style: solid;
    border-radius: ${theme.spacings[borderRadius]};
    border-color: ${theme.colors[borderColor][borderShade]};

    background-color: ${backgroundColor
      ? theme.colors[backgroundColor][backgroundShade]
      : "transparent"};
    box-shadow: ${theme.shadows[shadow]};
    row-gap: ${theme.spacings[rowGap]};
  `}
`;

export default Box;
