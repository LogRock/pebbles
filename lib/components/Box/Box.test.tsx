import React from "react";
import Box from ".";
import { render } from "../../utils/test-utils";

describe("Box", () => {
  it("renders with default style", () => {
    const { container } = render(<Box />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders with custom styles", () => {
    const { container } = render(
      <Box
        align="end"
        margin="xxbig"
        padding="xlg"
        backgroundColor="destructive"
        backgroundShade="100"
        borderRadius="lg"
        shadow="upMedium"
        direction="row"
        rowGap="lg"
        colGap="huge"
        justify="space-between"
        noWrap
      />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
