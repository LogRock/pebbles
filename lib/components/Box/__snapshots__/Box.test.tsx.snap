// Vitest Snapshot v1

exports[`Box > renders with custom styles 1`] = `
<div
  class="sc-aXZVg dnhOFF"
  direction="row"
/>
`;

exports[`Box > renders with default style 1`] = `
<div
  class="sc-aXZVg iviBaG"
/>
`;
